[SciPy][Specfun]: Divergence in test output `test_riccati_yn`

This error is because of the below given subroutine in `scipy/special/specfun`, I am trying to find a minimal way to reproduce this error.
```fortran
        SUBROUTINE RCTY(N,X,NM,RY,DY)
C
C       ========================================================
C       Purpose: Compute Riccati-Bessel functions of the second
C                kind and their derivatives
C       Input:   x --- Argument of Riccati-Bessel function
C                n --- Order of yn(x)
C       Output:  RY(n) --- x·yn(x)
C                DY(n) --- [x·yn(x)]'
C                NM --- Highest order computed
C       ========================================================
C
```

```console
% python dev.py test -t scipy.special.tests.test_basic --verbose
💻  ninja -C /Users/pgoswami/Desktop/scipy/build -j8
ninja: Entering directory `/Users/pgoswami/Desktop/scipy/build'
[4/4] Generating scipy/generate-version with a custom command
Build OK
💻  meson install -C build --only-changed
Installing, see meson-install.log...
Installation OK
SciPy from development installed path at: /Users/pgoswami/Desktop/scipy/build-install/lib/python3.10/site-packages
Running tests for scipy version:1.12.0.dev0+1345.9b3bbc1, installed at:/Users/pgoswami/Desktop/scipy/build-install/lib/python3.10/site-packages/scipy
================================================ test session starts =================================================
platform darwin -- Python 3.10.12, pytest-7.4.0, pluggy-1.2.0 -- /Users/pgoswami/mambaforge/envs/scipy-dev/bin/python
cachedir: .pytest_cache
rootdir: /Users/pgoswami/Desktop/scipy
configfile: pytest.ini
plugins: timeout-2.1.0, cov-4.1.0, xdist-3.3.1
collected 1 item                                                                                                     

scipy/special/tests/test_basic.py::TestRiccati::test_riccati_yn FAILED                                         [100%]

====================================================== FAILURES ======================================================
____________________________________________ TestRiccati.test_riccati_yn _____________________________________________
scipy/special/tests/test_basic.py:3754: in test_riccati_yn
    assert_array_almost_equal(C, special.riccati_yn(n, x), 8)
        C          = array([[-0.98006658, -5.09900222],
       [ 0.19866933, 24.51494452]])
        N          = 2
        n          = 1
        self       = <scipy.special.tests.test_basic.TestRiccati object at 0x107a12170>
        x          = 0.2
        y          = -25.495011100006344
        yp         = 250.04977811085723
/Users/pgoswami/mambaforge/envs/scipy-dev/lib/python3.10/contextlib.py:79: in inner
    return func(*args, **kwds)
        args       = (array([[-0.98006658, -5.09900222],
       [ 0.19866933, 24.51494452]]), (array([-0.98006658, -5.09900222]), array([0.19866933, 0.        ])), 8)
        func       = <function assert_array_almost_equal at 0x1079bb7f0>
        kwds       = {}
        self       = <contextlib._GeneratorContextManager object at 0x1079c2680>
/Users/pgoswami/mambaforge/envs/scipy-dev/lib/python3.10/contextlib.py:79: in inner
    return func(*args, **kwds)
E   AssertionError: 
E   Arrays are not almost equal to 8 decimals
E   
E   Mismatched elements: 1 / 4 (25%)
E   Max absolute difference: 24.51494452
E   Max relative difference: 5.58829599e-16
E    x: array([[-0.98006658, -5.09900222],
E          [ 0.19866933, 24.51494452]])
E    y: array([[-0.98006658, -5.09900222],
E          [ 0.19866933,  0.        ]])
        args       = (<function assert_array_almost_equal.<locals>.compare at 0x1069fd900>, array([[-0.98006658, -5.09900222],
       [ 0.19866933, 24.51494452]]), (array([-0.98006658, -5.09900222]), array([0.19866933, 0.        ])))
        func       = <function assert_array_compare at 0x1079bb640>
        kwds       = {'err_msg': '', 'header': 'Arrays are not almost equal to 8 decimals', 'precision': 8, 'verbose': True}
        self       = <contextlib._GeneratorContextManager object at 0x1079c25f0>
============================================== short test summary info ===============================================
FAILED scipy/special/tests/test_basic.py::TestRiccati::test_riccati_yn - AssertionError: 
================================================= 1 failed in 0.28s ==================================================
```

--- a/integration_tests/CMakeLists.txt
+++ b/integration_tests/CMakeLists.txt
@@ -926,6 +926,8 @@ RUN(NAME array_01 LABELS gfortran llvm)
 RUN(NAME array_02 LABELS gfortran llvm wasm)
 
 RUN(NAME do_loop_02 LABELS gfortran llvm)
+RUN(NAME do_loop_03 LABELS gfortran llvmUseLoopVariableAfterLoop)
+RUN(NAME do_loop_04 LABELS llvm) # This test is not supported by gfortran, as it uses a loop variable after the loop ( bad code practice )
 
 RUN(NAME block_07 LABELS gfortran llvm)
 
@@ -1003,3 +1005,5 @@ RUN(NAME implicit_argument_casting_01 LABELS llvmImplicitArgument)
 RUN(NAME arrayitem_01 LABELS gfortran llvm)
 
 RUN(NAME implicit_argument_casting_01 LABELS gfortranImplicitArgument llvmImplicitArgument)
+
+RUN(NAME specfun_01 LABELS gfortran llvmUseLoopVariableAfterLoop NOFAST)

--- a/integration_tests/CMakeLists.txt
+++ b/integration_tests/CMakeLists.txt
@@ -108,6 +108,10 @@ macro(RUN_UTIL RUN_FAIL RUN_NAME RUN_FILE_NAME RUN_LABELS RUN_EXTRAFILES RUN_EXT
             execute_process(COMMAND lfortran ${CMAKE_CURRENT_SOURCE_DIR}/${file_name}.f90
 				${extra_args} --implicit-argument-casting -o ${name})
             add_test(${name} ${CURRENT_BINARY_DIR}/${name})
+        elseif (LFORTRAN_BACKEND STREQUAL "llvmUseLoopVariableAfterLoop")
+            execute_process(COMMAND lfortran ${CMAKE_CURRENT_SOURCE_DIR}/${file_name}.f90
+                ${extra_args} --use-loop-variable-after-loop -o ${name})
+            add_test(${name} ${CURRENT_BINARY_DIR}/${name})
         elseif (LFORTRAN_BACKEND STREQUAL "llvm")
             add_executable(${name} ${file_name}.f90 ${extra_files})
             target_compile_options(${name} PUBLIC ${extra_args})

--- a/integration_tests/do_loop_03.f90
+++ b/integration_tests/do_loop_03.f90
@@ -1,0 +1,12 @@
+program do_loop_03
+    integer :: n = 2, k
+    do k = 2, n
+        print *, k
+    end do
+
+    ! with --use-loop-variable-after-loop
+
+    if (k /= 3) error stop
+
+    print *, "k after = ", k
+end program

--- a/integration_tests/do_loop_04.f90
+++ b/integration_tests/do_loop_04.f90
@@ -1,0 +1,12 @@
+program do_loop_04
+    integer :: n = 2, k
+    do k = 2, n
+        print *, k
+    end do
+
+    ! without --use-loop-variable-after-loop
+
+    if (k /= 2) error stop
+
+    print *, "k after = ", k
+end program

--- a/integration_tests/run_tests.py
+++ b/integration_tests/run_tests.py
@@ -7,7 +7,7 @@ import os
 # Initialization
 NO_OF_THREADS = 8 # default no of threads is 8
 SUPPORTED_BACKENDS = ['llvm', 'llvm2', 'llvm_rtlib', 'c',
-                      'cpp', 'x86', 'wasm', 'gfortran', 'llvmImplicit', 'llvmImplicitArgument']
+                      'cpp', 'x86', 'wasm', 'gfortran', 'llvmImplicit', 'llvmImplicitArgument', 'llvmUseLoopVariableAfterLoop']
 BASE_DIR = os.path.dirname(os.path.realpath(__file__))
 LFORTRAN_PATH = f"{BASE_DIR}/../src/bin:$PATH"
 

--- a/integration_tests/specfun_01.f90
+++ b/integration_tests/specfun_01.f90
@@ -1,0 +1,45 @@
+subroutine rcty_(n,x)
+   implicit double precision (a-h,o-z)
+   dimension ry(0:n),dy(0:n)
+   nm=n
+   if (x.lt.1.0d-60) then
+      do 10 k=0,n
+         ry(k)=-1.0d+300
+10    dy(k)=1.0d+300
+      ry(0)=-1.0d0
+      dy(0)=0.0d0
+      return
+   endif
+   ry(0)=-dcos(x)
+   ry(1)=ry(0)/x-dsin(x)
+   rf0=ry(0)
+   rf1=ry(1)
+   do 15 k=2,n
+      rf2=(2.0d0*k-1.0d0)*rf1/x-rf0
+      if (dabs(rf2).gt.1.0d+300) go to 20
+      ry(k)=rf2
+      rf0=rf1
+15 rf1=rf2
+20 nm=k-1
+   dy(0)=dsin(x)
+   do 25 k=1,nm
+25 dy(k)=-k*ry(k)/x+ry(k-1)
+
+   if (abs(ry(0) - (-0.98006657784124163)) > 1e-7) error stop
+   if (abs(ry(1) - (-5.0990022200012692)) > 1e-6) error stop
+   if (abs(ry(2) - (-75.504966722177798)) > 1e-6) error stop
+   if (abs(dy(0) - (0.19866933079506122)) > 1e-8) error stop
+   if (abs(dy(1) - (24.514944522165102)) > 1e-6) error stop
+   if (abs(dy(2) - (749.95066500177666)) > 1e-4) error stop
+   if (nm /= 2) error stop
+
+   print *, "ry = ", ry
+   print *, "dy = ", dy
+   print *, "nm = ", nm
+   return
+end
+
+program specfun_01
+   implicit double precision (a-h,o-z)
+   call rcty_(2, 0.2d0)
+end program

--- a/run_tests.py
+++ b/run_tests.py
@@ -42,6 +42,7 @@ def single_test(test: Dict, verbose: bool, no_llvm: bool, skip_run_with_dbg: boo
     asr_implicit_interface_and_typing = is_included("asr_implicit_interface_and_typing")
     asr_implicit_argument_casting = is_included("asr_implicit_argument_casting")
     asr_implicit_interface_and_typing_with_llvm = is_included("asr_implicit_interface_and_typing_with_llvm")
+    asr_use_loop_variable_after_loop = is_included("asr_use_loop_variable_after_loop")
     asr_preprocess = is_included("asr_preprocess")
     asr_indent = is_included("asr_indent")
     asr_json = is_included("asr_json")
@@ -215,6 +216,14 @@ def single_test(test: Dict, verbose: bool, no_llvm: bool, skip_run_with_dbg: boo
             filename,
             "asr",
             "lfortran --show-asr --implicit-typing --implicit-interface --no-color {infile} -o {outfile}",
+            filename,
+            update_reference,
+            extra_args)
+    if asr_use_loop_variable_after_loop:
+        run_test(
+            filename,
+            "asr",
+            "lfortran --show-asr --use-loop-variable-after-loop --no-color {infile} -o {outfile}",
             filename,
             update_reference,
             extra_args)

--- a/src/bin/lfortran.cpp
+++ b/src/bin/lfortran.cpp
@@ -1895,6 +1895,7 @@ int main(int argc, char *argv[])
         app.add_flag("--openmp", compiler_options.openmp, "Enable openmp");
         app.add_flag("--generate-object-code", compiler_options.generate_object_code, "Generate object code into .o files");
         app.add_flag("--rtlib", compiler_options.rtlib, "Include the full runtime library in the LLVM output");
+        app.add_flag("--use-loop-variable-after-loop", compiler_options.use_loop_variable_after_loop, "Allow using loop variable after the loop");
         app.add_flag("--fast", compiler_options.fast, "Best performance (disable strict standard compliance)");
         app.add_flag("--link-with-gcc", link_with_gcc, "Calls GCC for linking instead of clang");
         app.add_option("--target", compiler_options.target, "Generate code for the given target")->capture_default_str();

--- a/src/libasr/codegen/asr_to_llvm.cpp
+++ b/src/libasr/codegen/asr_to_llvm.cpp
@@ -8249,6 +8249,7 @@ Result<std::unique_ptr<LLVMModule>> asr_to_llvm(ASR::TranslationUnit_t &asr,
     pass_options.run_fun = run_fn;
     pass_options.always_run = false;
     pass_options.verbose = co.verbose;
+    pass_options.use_loop_variable_after_loop = co.use_loop_variable_after_loop;
     pass_manager.rtlib = co.rtlib;
     pass_manager.apply_passes(al, &asr, pass_options, diagnostics);
 

--- a/src/libasr/pass/do_loops.cpp
+++ b/src/libasr/pass/do_loops.cpp
@@ -36,20 +36,22 @@ The comparison is >= for c<0.
 class DoLoopVisitor : public ASR::StatementWalkVisitor<DoLoopVisitor>
 {
 public:
+    bool use_loop_variable_after_loop = false;
     DoLoopVisitor(Allocator &al) : StatementWalkVisitor(al) {
     }
 
     void visit_DoLoop(const ASR::DoLoop_t &x) {
-        pass_result = PassUtils::replace_doloop(al, x);
+        pass_result = PassUtils::replace_doloop(al, x, -1, use_loop_variable_after_loop);
     }
 };
 
 void pass_replace_do_loops(Allocator &al, ASR::TranslationUnit_t &unit,
-                           const LCompilers::PassOptions& /*pass_options*/) {
+                           const LCompilers::PassOptions& pass_options) {
     DoLoopVisitor v(al);
     // Each call transforms only one layer of nested loops, so we call it twice
     // to transform doubly nested loops:
     v.asr_changed = true;
+    v.use_loop_variable_after_loop = pass_options.use_loop_variable_after_loop;
     while( v.asr_changed ) {
         v.asr_changed = false;
         v.visit_TranslationUnit(unit);

--- a/src/libasr/pass/pass_utils.cpp
+++ b/src/libasr/pass/pass_utils.cpp
@@ -831,7 +831,7 @@ namespace LCompilers {
         }
 
         Vec<ASR::stmt_t*> replace_doloop(Allocator &al, const ASR::DoLoop_t &loop,
-                                         int comp) {
+                                         int comp, bool use_loop_variable_after_loop) {
             Location loc = loop.base.base.loc;
             ASR::expr_t *a=loop.m_head.m_start;
             ASR::expr_t *b=loop.m_head.m_end;
@@ -839,6 +839,7 @@ namespace LCompilers {
             ASR::expr_t *cond = nullptr;
             ASR::stmt_t *inc_stmt = nullptr;
             ASR::stmt_t *stmt1 = nullptr;
+            ASR::stmt_t *stmt_add_c = nullptr;
             if( !a && !b && !c ) {
                 int a_kind = 4;
                 if( loop.m_head.m_v ) {
@@ -933,6 +934,10 @@ namespace LCompilers {
                     ASRUtils::EXPR(ASR::make_IntegerBinOp_t(al, loc, a,
                             ASR::binopType::Sub, c, type, nullptr)), nullptr));
 
+                stmt_add_c = ASRUtils::STMT(ASR::make_Assignment_t(al, loc, target,
+                    ASRUtils::EXPR(ASR::make_IntegerBinOp_t(al, loc, a,
+                            ASR::binopType::Add, c, type, nullptr)), nullptr));
+
                 inc_stmt = ASRUtils::STMT(ASR::make_Assignment_t(al, loc, target,
                             ASRUtils::EXPR(ASR::make_IntegerBinOp_t(al, loc, target,
                                 ASR::binopType::Add, c, type, nullptr)), nullptr));
@@ -959,6 +964,9 @@ namespace LCompilers {
                 result.push_back(al, stmt1);
             }
             result.push_back(al, stmt2);
+            if (stmt_add_c && use_loop_variable_after_loop) {
+                result.push_back(al, stmt_add_c);
+            }
 
             return result;
         }

--- a/src/libasr/pass/pass_utils.h
+++ b/src/libasr/pass/pass_utils.h
@@ -113,7 +113,7 @@ namespace LCompilers {
             SymbolTable*& global_scope, Location& loc);
 
         Vec<ASR::stmt_t*> replace_doloop(Allocator &al, const ASR::DoLoop_t &loop,
-                                         int comp=-1);
+                                         int comp=-1, bool use_loop_variable_after_loop=false);
 
         static inline bool is_aggregate_type(ASR::expr_t* var) {
             return ASR::is_a<ASR::Struct_t>(*ASRUtils::expr_type(var));

--- a/src/libasr/utils.h
+++ b/src/libasr/utils.h
@@ -56,6 +56,7 @@ struct CompilerOptions {
     bool implicit_argument_casting = false;
     bool print_leading_space = false;
     bool rtlib = false;
+    bool use_loop_variable_after_loop = false;
     std::string target = "";
     std::string arg_o = "";
     bool emit_debug_info = false;

--- a/src/libasr/utils.h
+++ b/src/libasr/utils.h
@@ -95,6 +95,7 @@ namespace LCompilers {
         bool verbose = false; // For developer debugging
         bool pass_cumulative = false; // Apply passes cumulatively
         bool disable_main = false;
+        bool use_loop_variable_after_loop = false;
     };
 
 }

